lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2974,6 +3009,25 @@ ConformanceChecker::inferTypeWitnessesViaValueWitnesses(ValueDecl *req) {
                                          result.second->getCanonicalType()})
                              .second)
                          return true;
+                       
+                       // Filter out circular possibilities, e.g. that
+                       // AssocType == S.AssocType or
+                       // AssocType == Foo<S.AssocType>.
+                       bool containsTautologicalType =
+                         result.second.findIf([&](Type t) -> bool {
+                           if (auto dmt = t->getAs<DependentMemberType>()) {
+                             if (associatedTypesAreSameEquivalenceClass(
```

Comment:
I think you should also check equivalence of the base; this means we we'll throw away, e.g., `Self.B := Self.A.B` early on, which we probably don't want.